contracts/protocol/tokenization/NTokenUniswapV3.sol

```diff
@@ -107,44 +109,73 @@ contract NTokenUniswapV3 is NToken {
             );
         }
 
+        (
+            ,
+            ,
+            address token0,
+            address token1,
+            ,
+            ,
+            ,
+            ,
+            ,
+            ,
+            ,
+
+        ) = INonfungiblePositionManager(_underlyingAsset).positions(tokenId);
+
+        address weth = _addressesProvider.getWETH();
+        receiveEthAsWeth = (receiveEthAsWeth &&
+            (token0 == weth || token1 == weth));
+
         INonfungiblePositionManager.CollectParams
             memory collectParams = INonfungiblePositionManager.CollectParams({
                 tokenId: tokenId,
-                recipient: _msgSender(),
+                recipient: receiveEthAsWeth ? address(this) : _msgSender(),
                 amount0Max: 2**128 - 1,
                 amount1Max: 2**128 - 1
             });
 
         (amount0, amount1) = INonfungiblePositionManager(_underlyingAsset)
             .collect(collectParams);
+
+        require(amount0 >= amount0Min, "Insufficient amount0");
+        require(amount1 >= amount1Min, "Insufficient amount1");
+
+        if (receiveEthAsWeth) {
+            uint256 balanceWeth = IERC20(weth).balanceOf(address(this));
+            if (balanceWeth > 0) {
+                IWETH(weth).withdraw(balanceWeth);
+                (bool success, ) = _msgSender().call{value: balanceWeth}(
```

Comment:
re-entrancy guard will be needed, is `gas limit` necessary?

Reply:
I think in our code base we have also `safeTransfer` for ETH, can we reuse?

Reply:
I think we don't need re-entrancy guard here because UniswapV3Pool control the position state and there is no erc20 asset on this contract. NonfungiblePositionManager also don't have any re-entrancy guard.